bukkit/src/main/java/dev/jsinco/brewery/bukkit/listener/InventoryEventListener.java

```diff
@@ -0,0 +1,342 @@
+            if (!transactionEvent.callEvent()) {
+                if (transactionEvent.getCancelState() instanceof CancelState.PermissionDenied(Component denyMessage)) {
+                    event.getWhoClicked().sendMessage(denyMessage);
+                }
+                event.setResult(Event.Result.DENY);
+                return;
+            }
+        }
+        Bukkit.getScheduler().runTask(TheBrewingProject.getInstance(), () -> {
+            for (ItemTransactionEvent<?> transactionEvent : transactions) {
+                ItemTransactionSession<?> session = transactionEvent.getTransactionSession();
+                ItemTransaction transaction = session.getTransaction();
+                ItemStack itemStack = session.getResult() == null ? null : session.getResult().get();
+                if (transaction.to() instanceof ItemTransaction.Cursor
+                        && transaction.itemStack().equals(event.getView().getCursor())
+                ) {
+                    event.getView().setCursor(itemStack);
+                }
+                if (transaction.to() instanceof ItemTransaction.RawPosition(int pos)
+                        && transaction.itemStack().equals(event.getView().getItem(pos))
+                ) {
+                    event.getView().setItem(pos, itemStack);
+                }
+                if (transaction.to() instanceof ItemTransaction.UpperInventoryPosition(int pos)
+                        && transaction.itemStack().equals(event.getView().getTopInventory().getItem(pos))
+                ) {
+                    event.getView().getTopInventory().setItem(pos, itemStack);
+                }
+                if (transaction.to() instanceof ItemTransaction.LowerInventoryPosition(int pos)
+                        && transaction.itemStack().equals(event.getView().getBottomInventory().getItem(pos))
+                ) {
+                    event.getView().getBottomInventory().setItem(pos, itemStack);
+                }
+            }
+        });
+    }
+
+    private List<? extends ItemTransactionEvent<?>> compileTransactionsFromClick(InventoryClickEvent event, boolean upperInventoryIsClicked,
+                                                                                 InventoryAccessible<ItemStack, Inventory> inventoryAccessible) {
+        Player player = event.getWhoClicked() instanceof Player temp ? temp : null;
+        if (event.getAction() == InventoryAction.MOVE_TO_OTHER_INVENTORY) {
+            if (upperInventoryIsClicked) {
+                return findEmptyPositions(event.getView(), event.getCurrentItem(), false)
+                        .stream()
+                        .map(inventoryPosition -> eventFromStructure(
+                                inventoryAccessible,
+                                new ItemTransaction.RawPosition(event.getRawSlot()),
+                                inventoryPosition,
+                                event.getCurrentItem(),
```

Comment:
The `event.getCurrentItem()` and other items referenced by the event are mapped to real items in memory (either items or slots, Spigot docs aren't clear on this), so Spigot could modify the item before the next tick. Is this properly accounted for in the scheduled task? If not, we may need to clone ItemStacks *before* passing them to `ItemSource`.